### PR TITLE
support pushing if down to kv

### PR DESCRIPTION
**Pull request by XuHuaiyu**
This commit pushes `if` down to kv.
PTAL @shenli @hanfei1991 @coocood @zimulala @tiancaiamao

### CODE REVIEW

File: plan/expr_to_pb.go [hunk 1]
```diff
@@ -156,27 +156,32 @@ func compareOpsToPBExpr(client kv.Client, expr *expression.ScalarFunction) *tipb
...
 				return nil
-			case '%':
-				if i != 0 && i != len(pattern.Value.GetString())-1 {
-					return nil
-				}
 			}
 		}
-		tp = tipb.ExprType_Like
 	}
+	tp := tipb.ExprType_Like
```
Comment (shenli) [inline plan/expr_to_pb.go]: Move this check to the beginning of this function.

**Comment by XuHuaiyu** [comment]
PTAL @shenli

File: distsql/xeval/eval_control_funcs.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+func (e *Evaluator) evalControlFuncs(expr *tipb.Expr) (d types.Datum, err error) {
+	switch expr.GetTp() {
+	case tipb.ExprType_Case:
+		return e.evalCaseWhen(expr)
```
Comment (hanfei1991) [inline distsql/xeval/eval_control_funcs.go]: move eval case when to this file ?

File: plan/expr_to_pb.go [hunk 3]
```diff
@@ -428,13 +363,37 @@ func aggFuncToPBExpr(client kv.Client, aggFunc expression.AggregationFunction) *
 }
 
 func builtinFuncToPBExpr(client kv.Client, expr *expression.ScalarFunction) *tipb.Expr {
+	switch expr.FuncName.L {
```
Comment (hanfei1991) [inline plan/expr_to_pb.go]: why not move this logic to line 133?

Comment (XuHuaiyu) [inline plan/expr_to_pb.go]: I think that if we move the switch-case to line133, `scalarFuncToPBExpr` may be very long if we support more funcs in the future. @hanfei1991

**Comment by XuHuaiyu** [comment]
@shenli @hanfei1991 PTAL

**Comment by hanfei1991** [comment]
LGTM

File: distsql/xeval/eval_control_funcs.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+func (e *Evaluator) evalIf(expr *tipb.Expr) (d types.Datum, err error) {
+	if len(expr.Children) != 3 {
+		err = ErrInvalid.Gen("IF needs 3 operands but got %d", len(expr.Children))
+		return
+	}
+	child1, err := e.Eval(expr.Children[0])
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	d, err = e.Eval(expr.Children[2])
```
Comment (shenli) [inline distsql/xeval/eval_control_funcs.go]: conditionTrue := false
        if !child1.IsNull() {
	x, err := child1.ToBool()
 	if err != nil {
 		return d, errors.Trace(err)
 	}
 	if x == 1 {
           conditionTrue = true
        }
     }
     if conditionTrue {
     // eval  and return expr.Children[1]
     } 
    // eval and return expr.Children[2]

**Comment by shenli** [comment]
LGTM

**Pull request merged**